Sources/__SwiftOnServer_org/Documentation.docc/2024/websockets-tutorial-using-swift-and-hummingbird/websockets-tutorial-using-swift-and-hummingbird.md

```diff
@@ -0,0 +1,455 @@
+
+WebSockets are particularly useful for applications that require low latency and high-frequency updates, such as online gaming, chat applications, and live data feeds. The protocol supports full-duplex communication, meaning data can be sent and received simultaneously. This efficient data transfer method helps in creating more interactive applications, providing a smoother user experience.
+
+## WebSockets vs long polling vs HTTP streaming, and server-sent events
+
+Similar to WebSockets, there are other methods of achieving real-time capabilities. Techniques like HTTP polling, HTTP streaming, Comet, and SSE (server-sent events) come to mind. Let’s explore how these methods differ.
+
+### Long polling (HTTP polling)
+
+Long polling (HTTP polling) was one of the first methods to address real-time data fetching. It involved the client repeatedly sending requests to the server. The server holds the request open until there’s new data or a timeout occurs. Once data is available, the server responds, and the client immediately sends a new request. However, long polling has several issues, including header overhead, latency, timeouts, and caching problems.
+
+### HTTP streaming
+
+HTTP streaming reduces network latency by keeping the initial request open indefinitely. Unlike long polling, the server does not close the connection after sending data; it keeps it open to send new updates whenever there is a change.
+
+### SSE - Server-sent events
+
+Server-sent events (SSE) allow the server to push data to the client, similar to HTTP streaming. SSE is a standardized version of HTTP streaming and includes a built-in browser API. However, SSE is not suitable for applications like chats or games that require two-way communication since it only allows a one-way data flow from the server to the client. SSE uses traditional HTTP and has limitations on the number of open connections.
+
+## Why to use WebSockets?
+
+The above methods are less efficient on a protocol level and often seem like workarounds to make a request-reply protocol appear full-duplex. WebSockets are designed for full-duplex communication, and are more lightweight than its alternatives.
+
+### Security (WSS)
+
+WebSocket (WS) uses a plain-text TCP connection, making it less secure and easy to intercept. WebSocket Secure (WSS), like HTTPS, uses TLS to protect the TCP connection. It's identical to WebSocket otherwise. WSS protects against man-in-the-middle attacks but does not offer cross-origin or application-level security. Developers should add URL origin checks and strong authentication. 
+
+## How to use WebSockets to build a real-time chat application?
+
+The [Hummingbird Websocket](https://github.com/hummingbird-project/hummingbird-websocket) library is an extension for the Hummingbird web framework. This library provides WebSocket Client- and Serve support using the latest structured concurrency features like ``TaskGroup`` and ``AsyncSequence``s.
```

Comment:
We should add a `<!-- TODO -->` comment so that we remember to build a link to the AsyncSequences article when that's ready